vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go

```diff
@@ -0,0 +1,208 @@
+// This is best used with the `maintainedRatio` policy or with a single controlled resource.
+type ReplicaRestrictionsPostProcessor struct {
+	ControllerFetcher controllerfetcher.ControllerFetcher
+}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *ReplicaRestrictionsPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, _ *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	watermarks := readWatermarksFromVPAAnnotations(vpa)
+	if watermarks == nil || len(watermarks) == 0 || recommendation == nil {
+		return recommendation
+	}
+	lowWatermark, highWatermark := watermarks[0], watermarks[1]
+
+	podTemplate, err := r.ControllerFetcher.GetPodTemplateFromTopMostWellKnown(&controllerfetcher.ControllerKeyWithAPIVersion{
+		ControllerKey: controllerfetcher.ControllerKey{
+			Namespace: vpa.ID.Namespace,
+			Kind:      vpa.TargetRef.Kind,
+			Name:      vpa.TargetRef.Name,
+		},
+		ApiVersion: vpa.TargetRef.APIVersion,
+	})
+	if err != nil {
+		klog.Errorf("Failed to apply ReplicaRestrictionsPostProcessor (controller fetch) to vpa %s/%s due to error: %#v", vpa.ID.Namespace, vpa.ID.VpaName, err)
+		return recommendation
+	}
+
+	pod := newPodFromTemplate(podTemplate)
+
+	updatedRecommendation, err := r.apply(recommendation, vpa.PodCount, lowWatermark, highWatermark, pod)
+	if err != nil {
+		klog.Errorf("Failed to apply ReplicaRestrictionsPostProcessor to vpa %s/%s due to error: %#v", vpa.ID.Namespace, vpa.ID.VpaName, err)
+	}
+	return updatedRecommendation
+}
+
+func readWatermarksFromVPAAnnotations(vpa *model.Vpa) []int {
+	var watermarks []int
+
+	for key, value := range vpa.Annotations {
+		if key != vpaPostProcessorPrefix + vpaPostProcessorReplicaRestrictionsSuffix {
+			continue
+		}
+
+		if value == "" {
+			watermarks = []int{10, 100}
+		} else if err := json.Unmarshal([]byte(value), &watermarks); err != nil {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format, error:%#v", value, vpa.ID.Namespace, vpa.ID.VpaName, err)
+		}
+		if len(watermarks) != 2 {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format", value, vpa.ID.Namespace, vpa.ID.VpaName)
+			continue
+		}
+		break // There can be only one.
+	}
+	return watermarks
+}
+
+// ReplicaRestrictionsPostProcessor must implement RecommendationProcessor
+var _ RecommendationPostProcessor = &ReplicaRestrictionsPostProcessor{}
+
+// Apply returns a recommendation for the given pod, adjusted to obey maintainedRatio policy
+func (r *ReplicaRestrictionsPostProcessor) apply(
+	podRecommendation *vpa_types.RecommendedPodResources,
+	replicas, lowWatermark, highWatermark int,
+	pod *apiv1.Pod) (*vpa_types.RecommendedPodResources, error) {
+
+	if podRecommendation == nil {
+		// If there is no recommendation let's skip that post-processor.
+		return nil, nil
+	}
```

Comment:
I think we can skip even earlier in the post processor

Reply:
done